erc777.k

```diff
@@ -69,140 +72,379 @@ module ERC777
+                                Number ,  // amount
+                                Data ,    // userData
+                                Data      // operatorData
+                              )
+
+  rule <k> #callSender ( Operator , From , To , Value , Data , OperatorData )
+        => Hook . tokensToSend ( Operator , From , To , Value , Data , OperatorData ) ... </k>
+       <id> From </id>
+       <senderHook> Hook </senderHook>
        requires Hook =/=Int 0
 
-  rule <k> #hookFrom ( Operator , From , To , Value , Data , OperatorData )
-        => #mayRevert ... </k>
-       <senderHooks> ... From |-> Hook ... </senderHooks>
+  rule <k> #callSender ( Operator , From , To , Value , Data , OperatorData ) => . ... </k>
+       <id> From </id>
+       <senderHook> Hook </senderHook>
        requires Hook ==Int 0
 
-  syntax Stmt ::= #hookTo ( Address , Address , Address , Number , Data , Data )
-
-  rule <k> #hookTo ( Operator , From , To , Value , Data , OperatorData )
-        => #call ( #this , Hook , Operator , From , To , Value , Data , OperatorData ) ... </k>
-       <receiverHooks> ... To |-> Hook ... </receiverHooks>
+  syntax Stmt ::= #callRecipent ( Address , // operator
+                                  Address , // from
+                                  Address , // to
+                                  Number ,  // amount
+                                  Data ,    // userData
+                                  Data ,    // operatorData
+                                  Bool      // preventLocking
+                                )
+
+  // If the recipient is a contract, which has not registered an ERC777TokensRecipient implementation; the token contract:
+  // * MUST revert if the tokensReceived hook is called from a mint or send call.
+  // * SHOULD accept if the tokensReceived hook is called from an ERC20 transfer or transferFrom call.
+  rule <k> #callRecipent ( Operator , From , To , Value , Data , OperatorData , _ )
+        => Hook . tokensReceived ( Operator , From , To , Value , Data , OperatorData ) ... </k>
+       <id> To </id>
+       <receiverHook> Hook </receiverHook>
        requires Hook =/=Int 0
 
-  rule <k> #hookTo ( Operator , From , To , Value , Data , OperatorData )
-        => #ite ( #isContractAddress ( To ) , #revert , #mayRevert ) ... </k>
-       <receiverHooks> ... To |-> Hook ... </receiverHooks>
-       requires Hook ==Int 0
+  rule <k> #callRecipent ( Operator , From , To , Value , Data , OperatorData , PreventLocking )
+        => #ite ( ISCONTRACT , revert , . ) ... </k>
+       <id> To </id>
+       <isContract> ISCONTRACT </isContract>
+       <receiverHook> Hook </receiverHook>
+       requires Hook ==Int 0 andBool PreventLocking
+
+  rule <k> #callRecipent ( Operator , From , To , Value , Data , OperatorData , PreventLocking )
+        => . ... </k>
+       <id> To </id>
+       <receiverHook> Hook </receiverHook>
+       requires Hook ==Int 0 andBool notBool PreventLocking
```

Comment:
why don't you merge two rules into one? 

also, `PreventLocking` doesn't sound good to me, when i saw this in the reference implementation.
how about something like `erc20compatiblemode`.

Reply:
I think `<ERC20CompatibleC>` cell can be used here instead of the `PreventLocking`.